### ISSUE TITLE
providing fixing documentation and/or links to helpful urls

### ISSUE THREAD

**Issue by OliverLeitner** [drwetter/testssl.sh#194]
hi guys

very nice toy you got there...

ive been playing with it, the one thing im missing though is good documentation links on howto mitigate those attacks from your website (https://testssl.sh)

i.e. searching for a working fix for the BREACH attack on google could be avoided if you link to a resource you verify as working.

**Comment by drwetter** [comment]
Thx for the flowers. I'd prefer calling it tool.

Links with further infos are generally a good idea. The output needs to take this into the account though and somebody has to implement the "further infos".

For breach mitigation I use for Apache this: https://community.qualys.com/message/20352#20352

**Comment by drwetter** [comment]
As you might have followed the discussion in #213 it's not easy. That was only the attack and the relevance, then people might have lots of servers: nginx, Apache, lighttpd, OLS, F5 BigIP, varnish, IIS 6, IIS 7, IIS *, Cisco, etc. 

It certainly would be a nice to have. So far nobody stepped fwd saying, I am willing to work on this. 
There are some resources in the web though which could be used  for major servers. Links to them here in the wiki would be enough for a start.

**Comment by Harinus** [comment]
A documentation on how to fix CVE on system XY with configuration Z sounds like impossible to me. The Qualys SSL Labs give another example of documentation with the goal to inform the User/Admin about the Vulnerability 
<table><tr class="tableRow">
                            <td class="tableLabel"><font color="red">POODLE (SSLv3)</font></td>
                <td class="tableCell">
                    <font color="red">(<a style="color: red" href="https://community.qualys.com/blogs/securitylabs/2014/10/15/ssl-3-is-dead-killed-by-the-poodle-attack">more info</a>)</font>
                    <span class="greySmall">disable SSL3</span>                </td>
                    </tr></table>
In conclusion, after reading the article you know you should disable SSL3 to be save from the POODLE Attack.
Another examle would be Breach: Conclusion: DisableTLS/SSL-level compression

<table><tr class="tableRow">
                            <td class="tableLabel"><font color="red">BREACH</font></td>
                <td class="tableCell">
                    <font color="red">(<a style="color: red" href="http://breachattack.com/">more info</a>)</font>
                    <span class="greySmall">DisableTLS/SSL-level compression</span>                </td>
                    </tr></table>
How to do it in any particular case should be a google search executed by the admin who is responsible for fixing the vulnerability.
If you agree with this kind of documentation. I may invest some more time in finding, maybe exsting collections of such links, or generating one for this project by combining the Qualys SSL Labs data and other sources if required.

**Comment by drwetter** [comment]
Thx, @Harinus. I feel the same. 

Without sleeping over it: anybody could google for the fix and it doesn't have to be duplicated what others did.

OTOH this is typical pentester perspective ("oh, you have a problem here, go figure how you solve it") \</sarcasm\>. Also I got personal requests like that and from a perspectives from someone who doesn't have a solid background to find the right mitigation or to tell which is the right one it would -- maybe -- make sense to e.g. have a link collection.

I guess before somebody like you jumps in and fills that gap it would make sense to discuss how we could implement that and then see whether it makes sense at all. 

What you suggested sounds a reasonable approach. But where is the place in the output of testssl.sh and where to put the documentation with handy links?

**Comment by Harinus** [comment]
@drwetter i agree with you.. the pentester probably is happy with the CVE number only.. His reaction might be: Ah, this flaw.. again.. i need to to this or that to fix it. The documentation is an approach to bring this project closer to not so sophisticated Users, which i highly appreciate! Anyways if this person needs more information than "Set this or disable that" to guarantee security he should not be responsible for security issues ;) 

Technical aspect:
![unbenannt](https://cloud.githubusercontent.com/assets/5645416/11000581/f0955bce-84a2-11e5-87c9-dca8e4eb6afe.PNG)
Maybe as a link behind the CVE Number? Easy in HTML but is bash capable of this? I don't know
The collection of the links also in a section of the Readme?

**Comment by drwetter** [comment]
> Maybe as a link behind the CVE Number? Easy in HTML but is bash capable of this? 

The thing with that paragraph is that it is too wide anyway. ;-/ 

Links are not possible in text mode. Terminals like KDE's konsole recognize links and you can click on them. And they have to be real links -- otherwise they are not clickable. But I don't like solutions which are not portable.

Maybe a reference to a footnote like [1] in text mode and on the bottom the footnote with the real link? Also I could think of here doing the footnote only when there is a finding.

> The collection of the links also in a section of the Readme?

Additionally, but prefer not only. 

The Readme was up to yesterday more a description of what's going on recently on github and despite its name some people don't read it. ;-) -- at least not when it comes to look for a solution.

Where do we put the target of the links to? Is the wiki of github a good thing and can one link to different sections? Just wondering because I never worked with it and I experienced other thingies containing a wiki which is fairly limited (e.g. redmine).

Cheers, Dirk

**Comment by Harinus** [comment]
I will come back to this on monday.. There are some ideas on my mind i need to check for feasibility. Have a nice weekend. 
Best Regads, Martin

**Comment by drwetter** [comment]
Am 11/06/2015 um 05:00 PM schrieb Martin Hoffmann:
> I will come back to this on monday.. There are some ideas on my mind i need to check for
> feasibility. Have a nice weekend.

thx, you too.

Cheers, Dirk

**Comment by Harinus** [comment]
Hi again,

something like this?: 

![unbenannt](https://cloud.githubusercontent.com/assets/5645416/11031700/a1d0cc18-86d6-11e5-9f43-2978775937f4.png)

The collection of useful links can than take place in your wiki (which already exits):
https://github.com/drwetter/testssl.sh/wiki

(Wiki documentation: https://help.github.com/articles/about-github-wikis/)
You could create a few pages:
1) About the project (As replace for the readme)
2) Manual (Issue #130)
3) Vulnerability Actions
-> This chapter contains the links to in a format like this:
<table><tbody><tr>
                            <td>POODLE (SSLv3)</td>
                <td>
                    (<a href="https://community.qualys.com/blogs/securitylabs/2014/10/15/ssl-3-is-dead-killed-by-the-poodle-attack">more info</a>)
                    disable SSL3                </td>
                    </tr>
<tr>
                            <td>BREACH</td>
                <td>
                    (<a href="http://breachattack.com/">more info</a>)
                    DisableTLS/SSL-level compression                </td>
                    </tr></tbody></table>

Target of these links may be external sources (e.g. breachattack.com or own articles in the wiki if someone is willing to provide useful content.

**Comment by PeterMosmans** [comment]
Hi all,

Interesting idea and discussion!
To dump my 0.02 BTC of ideas here: Personally I think that the tool itself should *not* be chock-full of information/links on how to mitigate issues. This sounds like feature-creep which adds bloat to the script.
Just referring people to e.g. testssl.sh/mitigations or something like that with a single one-liner instead of actions per item sounds in my opinion more reasonable.
Again, the idea of referring people sounds great, but not per item/too much 'logic' in the script itself.

testssl.sh is pretty big as it stands, and I'd rather have it excel at correctly scanning TLS/SSL issues and reporting on vulnerabilities, than offering people detailed suggestions.

(but again, just my 0.02 BTC...)


Peter

**Comment by drwetter** [comment]
Hi,

this discussion became somewhat stale. I do not think there should be more than a link for further reading or a very small hint of description. Also maybe it's a good idea to have a cmd-line switch of either suppressing this additional screen output or adding that info. Off the top of my hat: I would vote for the latter as the output is already 16:9. Shouldn't become 21:9 ;-)

For JSON/CSV that can be handled differently.

Cheers, Dirk
